### PR TITLE
fix: decorators should now correctly respect account hoisting

### PR DESCRIPTION
**Pull request by moldy530**
# Pull Request Checklist

- [ ] Did you add new tests and confirm existing tests pass? (`yarn test`)
- [ ] Did you update relevant docs? (docs are found in the `site` folder, and guidelines for updating/adding docs can be found in the [contribution guide](https://github.com/alchemyplatform/aa-sdk/blob/main/CONTRIBUTING.md))
- [ ] Do your commits follow the [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/) standard?
- [ ] Does your PR title also follow the [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/) standard?
- [ ] If you have a breaking change, is it [correctly reflected in your commit message](https://www.conventionalcommits.org/en/v1.0.0/#examples)? (e.g. `feat!: breaking change`)
- [ ] Did you run lint (`yarn lint:check`) and fix any issues? (`yarn lint:write`)
- [ ] Did you follow the [contribution guidelines](https://github.com/alchemyplatform/aa-sdk/blob/main/CONTRIBUTING.md)?

---

## PR-Codex overview
This PR focuses on making changes related to the Smart Account Client in the codebase. 

### Detailed summary:
- Updated the `type` property of the `SmartAccountProvider` to `SmartAccountClient`.
- Added a new class `BaseError` that extends `ViemBaseError` and has a custom name and version.
- Updated the `AccountNotFoundError` class to extend `BaseError` and removed the `version` property.
- Updated the `createLightAccountAlchemyClient` function to extend `lightAccountClientActions`.
- Added a new class `IncompatibleClientError` that extends `BaseError`.
- Updated the `createAlchemySmartAccountClientFromRpcClient` function to return an object with additional properties and a new type.
- Updated the `install${contract.name}` method in the `management-actions.ts` file to throw an error if the client is not a `SmartAccountClient`.
- Updated various import statements in different files.
- Added a new function `isAlchemySmartAccountClient` to check if a client is an `AlchemySmartAccountClient`.
- Updated the `lightAccountClientActions` function to use the `Client` type instead of `SmartAccountClient`.
- Updated the `pluginManagerActions` function to use the `Client` type instead of `SmartAccountClient`.
- Updated the `signMessage` function to use the `Client` type instead of `BaseSmartAccountClient`.
- Updated the `signMessageWith6492` function to use the `Client` type instead of `BaseSmartAccountClient`.
- Updated the `getAddress` function to use the `Client` type instead of `BaseSmartAccountClient`.
- Updated the `signTypedData` function to use the `Client` type instead of `BaseSmartAccountClient`.
- Updated the `signTypedDataWith6492` function to use the `Client` type instead of `BaseSmartAccountClient`.
- Updated the `alchemyActions` function to use the `Client` type instead of `AlchemySmartAccountClient_Base`.
- Updated the `installPlugin` function to use the `Client` type instead of `SmartAccountClient` and added error handling for incompatible clients.
- Updated the import statements in the `installPlugin.ts` file.

> The following files were skipped due to too many changes: `packages/accounts/src/msca/plugin-manager/installPlugin.ts`, `packages/alchemy/src/client/smartAccountClient.ts`, `packages/accounts/src/msca/plugin-manager/uninstallPlugin.ts`, `packages/core/src/client/decorators/smartAccountClient.ts`, `packages/accounts/src/nani-account/transferNaniAccountOwnership.ts`, `packages/accounts/src/light-account/actions/transferOwnership.ts`, `packages/core/src/actions/smartAccount/checkGasSponsorshipEligibility.ts`, `packages/core/src/actions/smartAccount/waitForUserOperationTransacation.ts`, `site/.vitepress/sidebar/packages/aa-accounts.ts`, `packages/core/src/actions/smartAccount/buildUserOperationFromTxs.ts`, `packages/alchemy/src/actions/simulateUserOperationChanges.ts`, `packages/core/src/actions/smartAccount/buildUserOperation.ts`, `packages/core/src/actions/smartAccount/sendTransactions.ts`, `packages/core/src/actions/smartAccount/upgradeAccount.ts`, `packages/core/src/actions/smartAccount/dropAndReplaceUserOperation.ts`, `packages/core/src/actions/smartAccount/sendUserOperation.ts`, `packages/core/src/actions/smartAccount/buildUserOperationFromTx.ts`, `site/.vitepress/sidebar/packages/aa-ethers.ts`, `site/.vitepress/sidebar/packages/aa-alchemy.ts`, `packages/core/src/actions/smartAccount/sendTransaction.ts`, `packages/core/src/client/isSmartAccountClient.ts`, `packages/accounts/src/msca/account-loupe/decorator.ts`, `packages/accounts/plugingen/phases/plugin-actions/read-actions.ts`, `packages/accounts/plugingen/phases/plugin-actions/index.ts`, `packages/accounts/src/msca/plugins/multi-owner/plugin.ts`, `packages/accounts/src/msca/plugins/token-receiver/plugin.ts`, `packages/accounts/src/msca/plugins/session-key/plugin.ts`, `site/.vitepress/sidebar/packages/aa-signers.ts`, `site/.vitepress/sidebar/packages/aa-core.ts`, `packages/core/src/index.ts`, `site/.vitepress/config.ts`

> ✨ Ask PR-Codex anything about this PR by commenting with `/codex {your question}`

### CODE REVIEW

File: packages/core/src/index.ts
Comment (moldy530) [inline packages/core/src/index.ts:1]: this just reorders the dependencies alphabetically

**Comment by moldy530** [comment]
ah woops this wasn't meant to include the session key account deletion as well, but it was merged and approved in #395

File: packages/core/src/errors/base.ts [hunk 1]
```diff
@@ -0,0 +1,7 @@
+import { BaseError as ViemBaseError } from "viem";
+import { VERSION } from "../version.js";
+
+export class BaseError extends ViemBaseError {
+  override name = "AASDKError";
+  override version = VERSION;
+}
```
Comment (denniswon) [inline packages/core/src/errors/base.ts:7]: nice!

**Review by denniswon — APPROVED** [review]
I see what has changed, but can you explain before this change, why account hoisting was not respected, and this pr enables it to do?

**Review by denniswon — APPROVED** [review]
I see what has changed, but can you explain before this change, why account hoisting was not respected, and this pr enables it to do?

**Pull request merged**